### ISSUE TITLE
[HOLD for payment 2024-01-24] [$1000] Chat -  User experienced big delay when posting the messages

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21490]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging.new.expensify.com
2. Log in with own account
3. Open a public room https://staging.new.expensify.com/r/5593084223054221
4. Send messages

## Expected Result:
Messages arrived on time

## Actual Result:
User experienced big delay when posting the messages

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.29.0

**Reproducible in staging?:** y

**Reproducible in production?:** y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![image (83)](https://github.com/Expensify/App/assets/93399543/6bafdf6c-f8b0-4542-b2b6-4c8ebe61e5c9)

![image (82)](https://github.com/Expensify/App/assets/93399543/fe910d78-197b-4ac5-8db5-3e0a15d296e2)

**Expensify/Expensify Issue URL:**

**Issue reported by:** David Barrett

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686956800520439

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c5aa8e942309dcf4</li>
        <li>Upwork Job ID: 1673361704564097024</li>
        <li>Last Price Increase: 2023-06-26</li>
    </ul>
</details>

**Comment by quinthar** [comment]
This is pretty easily solved by just measuring the "skew" between the client and server clocks:

```
startTime = NOW();
response = send(GetCurrentTime);
endTime = NOW();
latency = (endTime - startTime)/2;
skew = response.serverTime - startTime + latency;
```

Then whenever we generate a client-side timestamp, we just add the skew and it will be a much closer approximation of what the server time is at that exact moment.

**Comment by quinthar** [comment]
Lots more detail on that Slack thread.

**Comment by lschurr** [comment]
Not overdue. Just reviewed and added External.

**Comment by arthurmfgtab** [comment]
Hey, @lschurr 
I'm Arthur from Callstack - expert contributor group - and I'd like to take a look at this issue

**Comment by lschurr** [comment]
Waiting on proposals.

**Comment by Piotrfj** [comment]
Hi I'm Piotr from Callstack - expert contributor group - will start investigation in this area

**Comment by Piotrfj** [comment]
Current experience of writing a message in desktop version:

https://github.com/Expensify/App/assets/41384305/a8773c80-b8e6-4ec5-bfc2-f4f59bf0fcd6

There is visible delay between clicking send icon (or pressing enter), but it is so small it makes me wonder if this is the the thing the issue is about.

If the issue is truly about that I think we could change the behavior in a way where the message is appearing instantly on the chat history and it will not wait for the response (because thats my guess of what is happening in this case).

Please provide me some clarification about that.

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Have you checked this [thread](https://app.slack.com/client/T03SC9DTT/C049HHMV9SM/thread/C049HHMV9SM-1686956800.520439). If not please read that, if you have any queries please add them to the thread, thanks!

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj any update?

**Comment by lschurr** [comment]
Are we still waiting on an update from [Piotrfj](https://github.com/Piotrfj)? @Santhosh-Sellavel?

**Comment by lschurr** [comment]
Bump @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
Yes we are waiting for @Piotrfj

**Comment by lschurr** [comment]
@Piotrfj - is this still being worked on?

**Comment by lschurr** [comment]
Just DM'd @Piotrfj in Slack. Do you have an update for us?

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Any update here?

**Comment by lschurr** [comment]
@NikkiWines - thoughts on how we should move forward here?

**Comment by NikkiWines** [comment]
Hmm, catching up since I've just gotten assigned. I think we should confirm that this is still a problem since it was originally reported over a month ago during a period of high volatility as we were pushing a lot of changes. From there we can decide how to move forward

**Comment by lschurr** [comment]
Chatted with @Piotrfj in Slack - they are going to try and replicate on Monday and will comment here.

**Comment by lschurr** [comment]
Bump again @Piotrfj

**Comment by lschurr** [comment]
Great, thanks! Assigned to you @arthurmfgtab

**Comment by arthurmfgtab** [comment]
Hey, @lschurr :)
I'm still getting up to speed on [this convo](https://expensify.slack.com/archives/C049HHMV9SM/p1686956800520439), I'll try to have something new tomorrow

**Comment by waterim** [comment]
Hey everyone, Im Artem from Callstack.
I'll take over on behalf of @arthurmfgtab

**Comment by lschurr** [comment]
Assigned to you @waterim

**Comment by waterim** [comment]
@lschurr Hello, started investigation, will update on Wednesday or Thursday(tomorrow we have a holiday) :)

**Comment by lschurr** [comment]
Bumping again @waterim to remove the Overdue status. Thanks!

**Comment by waterim** [comment]
@lschurr In progress :)

**Comment by waterim** [comment]
@lschurr I've got the issue, just one question, did someone manage to reproduce it locally for a better testing?

**Comment by lschurr** [comment]
@Santhosh-Sellavel were you able to reproduce locally?

**Comment by waterim** [comment]
@lschurr This one is a tricky, was able to reproduce after changing time locally, created a skew, testing now the solution

**Comment by waterim** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
User experienced big delay when posting the messages
### What is the root cause of that problem?
The main issue here is a clock sync. "created" field in optimistic data is taking a local machine time, but if the server clock is ahead of the user time, all messages from the backend will be displayed after your local message.
**Files we are working with**:
HttpUtils.js
ReportUtils.js
Report.js
Network.ts
### What changes do you think we should make in order to solve the problem?
In order to fix the clock sync issue we can add this part of the code to the [HttpUtils.js](https://github.com/Expensify/App/blob/64ff471fc614a57645edf34b5d2ac346046f12c1/src/libs/HttpUtils.js#L43)
As was discussed we can make this calculations only for “OpenApp”, “OpenReport” and "ReconnectApp"
```javascript
       const startTime = new Date();
       // here is fetch request ...
       const serverTime = new Date(response.headers.get('Date'));
       const endTime = new Date();
       const latency = (endTime.valueOf() - startTime.valueOf()) / 2;
       const skew = serverTime.valueOf() - startTime.valueOf() + latency;
       Onyx.merge(ONYXKEYS.NETWORK, {timeSkew: skew});
```
In Report and ReportUtils we are getting the networkTimeSkew from ONYXKEYS.NETWORK
And after all we need is to add the skew to the timestamp for the fields in optimistic data.
```
const timestamp = new Date().valueOf() + networkTimeSkew
DateUtils.getDBTime(timestamp)
```
In [ReportUtils](https://github.com/Expensify/App/blob/64ff471fc614a57645edf34b5d2ac346046f12c1/src/libs/ReportUtils.js#L1735) and [Report](https://github.com/Expensify/App/blob/64ff471fc614a57645edf34b5d2ac346046f12c1/src/libs/actions/Report.js#L255)
### What alternative solutions did you explore?
N/A

**Comment by lschurr** [comment]
Hi @zanyrenney - I'm heading OOO until Wed, Sept 6th so adding the BZ label again. The issue is assigned to @waterim who is working on a solution. I'll take this back from you when I return!

**Comment by zanyrenney** [comment]
sounds good!

**Comment by waterim** [comment]
@zanyrenney 
What do you think about my Proposal?

**Comment by Santhosh-Sellavel** [comment]
@waterim I'll look into it today, Sorry I missed it!

**Comment by Santhosh-Sellavel** [comment]
@waterim's [proposal](https://github.com/Expensify/App/issues/21490#issuecomment-1697412148) looks good to me!

Let's test it on PR!

C+ Reviewed
🎀 👀 🎀

**Comment by waterim** [comment]
@Santhosh-Sellavel Great, will create a PR :)

**Comment by Santhosh-Sellavel** [comment]
@waterim Any update here?

**Comment by waterim** [comment]
Opened a PR, everything is fine, but some tests are failing because of the skew, like time is wrong for some milliseconds, you can see it in the draft failing unit tests, thinking what I can do with that

**Comment by Santhosh-Sellavel** [comment]
@waterim 

I'm going OOO, once PR is ready for review. Request for a new C+ here.

ccL @lschurr

**Comment by waterim** [comment]
@Santhosh-Sellavel okay, thank you for the information.
I was OOO, callstack had a conference together with Expensify, will update on Monday.

**Comment by situchan** [comment]
I am interested in reviewing this as C+

**Comment by lschurr** [comment]
Removing @Santhosh-Sellavel - are you able to review the PR @situchan?

**Comment by situchan** [comment]
> re you able to review the PR @situchan?

yes

**Comment by lschurr** [comment]
Any update on this one @situchan?

**Comment by situchan** [comment]
PR is still in draft.
Latest update: https://github.com/Expensify/App/pull/26729#issuecomment-1715368348

**Comment by waterim** [comment]
Working now on this, hopefully will open tomorrow
@situchan if you have some ideas about failing tests, please feel free to add some comments
Sorry for a delay, had critical tasks

**Comment by waterim** [comment]
@situchan @lschurr Found a root, pushed change and opened a PR

**Comment by waterim** [comment]
@lschurr @situchan Do you have some updates?

**Comment by situchan** [comment]
updating today

**Comment by lschurr** [comment]
@situchan can you work on the checklist for this one?

**Comment by lschurr** [comment]
Looking for an update @waterim

**Comment by waterim** [comment]
Will take this on Monday, as previous PR was reverted, need to think why it was causing these issues!

**Comment by lschurr** [comment]
Not overdue. Sounds like @waterim will work on this today.

**Comment by lschurr** [comment]
Bump on this one @waterim

**Comment by lschurr** [comment]
Just clarifying @waterim - are you saying that the bug is fixed and we just need to issue payments and complete the [original checklist](https://github.com/Expensify/App/issues/21490#issuecomment-1745236839)?

**Comment by waterim** [comment]
No, I mean my previous solution, which was reverted works fine on localhost, and I cant reproduce it

**Comment by lschurr** [comment]
Ok, what are the next steps for this one @waterim @situchan?

**Comment by lschurr** [comment]
Bump @waterim @situchan

**Comment by situchan** [comment]
@waterim I think we should ask QA team to do regression test in new PR build so we can catch clear repro step and fix it.

**Comment by waterim** [comment]
@situchan, okay, will open a PR soon

**Comment by waterim** [comment]
@situchan @lschurr Reopened a PR

**Comment by lschurr** [comment]
What's the latest on this one @situchan @waterim?

**Comment by waterim** [comment]
This one was merged! 
Hopefully no regression will appear

**Comment by lschurr** [comment]
Looks like this didn't automatically move to Daily.

@situchan - Can you work through the checklist on this one?

**Comment by lschurr** [comment]
Payment summary:

- Bug report: Internal (no payment)
- Contributor: @waterim (contracted - no payment)
- Contributor+: @situchan $500 (to be paid in newdot)

**Comment by situchan** [comment]
BZ Checklist:
No PR caused regression.
This is edge case and I don't think we need regression test.

**Comment by situchan** [comment]
This is very old issue and can we keep original base price?

**Comment by situchan** [comment]
I am not eligible yet for newdot payment. still using upwork

**Comment by lschurr** [comment]
>This is very old issue and can we keep original base price?
There was a regression https://github.com/Expensify/App/pull/28673 which decreases the payment to $500

**Comment by situchan** [comment]
yes, there was regression but it couldn't have been caught during PR review. Both author and reviewer were not able to reproduce.
2nd PR was also QA'ed by 3rd party since again we're not able to reproduce.
We tried my best to find any regressions but unfortunately not.
And no other contributors were involved, so no overpayment for this issue.

Similar case: https://github.com/Expensify/App/issues/30042#issuecomment-1886914165

cc: @mountiny

**Comment by lschurr** [comment]
Thanks @situchan - we discussed in Slack and decided to keep the payment at $1000. I'll modify in Upwork.

**Comment by situchan** [comment]
Thanks so much

**Comment by lschurr** [comment]
New offer sent - can you accept in Upwork @situchan

**Comment by lschurr** [comment]
Great, payment sent in Upwork. We're all set.
